tools/scripts/add_library_tso_doublets.py

```diff
@@ -0,0 +1,180 @@
+    freq = (dist < dist_th) & (idx > adata[adata.obs["doublet_cell"] == False, :].shape[0])
+    score1 = freq.mean(axis=1)
+    score2 = freq[:, :np.int64(np.ceil(k/2))].mean(axis=1)
+    adata.obs["doublet_score"] = [np.max([score1[i], score2[i]]) for i in range(adata.shape[0])]   
+    doublet_csv=adata.obs.loc[~adata.obs_names.isin(doublet_obs_names), ["doublet_score"]]
+    
+    # Calculate the percentage of doublets with a doublet_score > 0.3
+    num_doublets = doublet_csv[doublet_csv["doublet_score"] > 0.3].shape[0]
+    total_cells = doublet_csv.shape[0]
+    percent_doublets = num_doublets / total_cells * 100
+
+    return doublet_csv, percent_doublets 
+
+
+# Function to calculate additional library metrics such as keeper cells
+def process_gex_data(gex_h5ad_modified, gex_nhash_id, library_csv, input_id, doublets, doublet_scores, counting_mode, expected_cells):
+    print("Reading Optimus h5ad:")
+    gex_data = gex_h5ad_modified
+    # NHashID is optional input, so the logic below sets it if undefined
+    if gex_nhash_id is not None:
+        gex_data.uns['NHashID'] = gex_nhash_id
+    else:
+        gex_nhash_id = "NA"
+        gex_data.uns['NHashID'] = gex_nhash_id
+
+    #gex_data.write(f"{input_id}.h5ad")
+
+    print("Reading library metrics")
+    library = pd.read_csv(library_csv, header=None)
+
+    # Calculates total library TSO metrics
+    # TSO reads refer to reads derived from the Template Switch Oligo
+    # TSO reads per cell are calculated from the BAM cN BAM tag
+    print("Calculating TSO frac")
+    tso_reads = gex_data.obs.tso_reads.sum() / gex_data.obs.n_reads.sum()
+    print("TSO reads:")
+    print(tso_reads)
+    
+    print("Calclating keeper metrics based on doublets and n_genes")
+    if counting_mode == "sc_rna":
+        gene_threshold = 1500
+    else:
+        gene_threshold = 1000
+
+    estimated_cells = len(gex_data[gex_data.obs["star_IsCell"]==True])
+    # Expected cells is the number expected from the experiment; usually 10,000 with 10x data
+    expected_cells = int(expected_cells)  # Placeholder, replace with actual value
+    
+    # Adding doublet scores to barcodes that have been called as cells
+    all_barcodes = pd.DataFrame(index=gex_data.obs_names)
+    # Merge doublet scores with all barcodes, filling missing values with NA
+    all_barcodes = all_barcodes.join(doublet_scores, how='left')
+    # Assign the doublet scores back to the adata object
+    gex_data.obs['doublet_score'] = all_barcodes['doublet_score']
+
+    # Adding keeper metrics
+    subset = gex_data[(gex_data.obs['star_IsCell']== True) & (gex_data.obs['doublet_score']<0.3) & (gex_data.obs['n_genes']> gene_threshold)]
+    keeper_cells = subset.shape[0]
+    keeper_mean_reads_per_cell = subset.obs["n_reads"].mean()
+    keeper_median_genes = subset.obs["n_genes"].median() 
+    percent_keeper = keeper_cells/estimated_cells
+    percent_usable = keeper_cells/expected_cells
+
+    # Updating library metrics
```

Comment:
consider using pd.concat() instead of manual dictionary creation: library = pd.concat([library, new_metrics])